### PR TITLE
Implement 1-RTT Exporter.

### PR DESCRIPTION
**Pull request by ekr**

### CODE REVIEW

**Comment by ekr** [comment]
@bifurcation: I think this is ready but I'd like a review from @martinthomson first.

File: server-state-machine.go [hunk 1]
```diff
@@ -785,6 +800,7 @@ type ServerStateWaitFinished struct {
 	handshakeHash       hash.Hash
 	clientTrafficSecret []byte
 	serverTrafficSecret []byte
+	exporterSecret      []byte
```
Comment (martinthomson) [inline server-state-machine.go]: The duplication here is starting to look pretty bad.  Any reason why these state variables can't be bound up into a struct that can be reused?  This is probably a question for @bifurcation who created this miss in the first place.

File: conn.go [hunk 2]
```diff
@@ -729,3 +727,20 @@ func (c *Conn) SendKeyUpdate(requestUpdate bool) error {
 func (c *Conn) GetHsState() string {
 	return reflect.TypeOf(c.hState).Name()
 }
+
+func (c *Conn) ComputeExporter(label string, context []byte, keyLength int) ([]byte, error) {
+	_, ok := c.hState.(StateConnected)
+	if !ok {
```
Comment (martinthomson) [inline conn.go]: s/ok/connected/, as above

File: conn.go [hunk 3]
```diff
@@ -729,3 +727,20 @@ func (c *Conn) SendKeyUpdate(requestUpdate bool) error {
 func (c *Conn) GetHsState() string {
 	return reflect.TypeOf(c.hState).Name()
 }
+
+func (c *Conn) ComputeExporter(label string, context []byte, keyLength int) ([]byte, error) {
```
Comment (martinthomson) [inline conn.go]: Just to confirm: the plan is to add a `ComputeEarlyExporter` function for the early exporter and then to factor out the necessary parts of this function.

File: conn.go (hunk 2 above)
Comment (martinthomson) [inline conn.go]: That said, if you want to factor out the early exporter, then checking that the appropriate secret is non-nil will achieve the same goal.

File: conn.go (hunk 3 above)
Comment (martinthomson) [inline conn.go]: I just learned something about go today.  Apparently `string` is [essentially the same](https://blog.golang.org/strings) as `[]byte`.  I don't know what to do about that though.

File: conn_test.go [hunk 4]
```diff
@@ -274,6 +274,12 @@ func assertKeySetEquals(t *testing.T, k1, k2 keySet) {
 	assertByteEquals(t, k1.key, k2.key)
 }
 
+func computeExporter(t *testing.T, c *Conn, label string, length int) []byte {
+	res, err := c.ComputeExporter(label, []byte{}, length)
```
Comment (martinthomson) [inline conn_test.go]: The tests should exercise non-zero-length contexts.

File: conn.go (hunk 3 above)
Comment (ekr) [inline conn.go]: yeah

File: conn_test.go (hunk 4 above)
Comment (ekr) [inline conn_test.go]: I didn't expose non-zero length contexts. Do you think we need them?

Comment (ekr) [inline conn_test.go]: Oh, never mind. Am idiot.

**Comment by bifurcation** [comment]
Replaced by #135

**Pull request closed without merging**